🚀 5단계 - 자동차 경주(리팩터링)

giyeon95 opened the pull request:
리뷰어님 안녕하세요!
마지막 미션 리뷰 요청 드립니다.

사실 크게 변경한 점이 없기도 하고.. 지난 번에 리뷰 주셨던 내용에 궁금한 점이 있어 추가 질문 드려요🙏

[CarStrategy 에서 바로 Engine을 리턴](https://github.com/next-step/kotlin-racingcar/pull/844#discussion_r1022969605)에 대해 주신 의견에서, 바로 Engine을 리턴하게 되면 다시 테스트하기 어려운 구조로 변하는 게 아닌가요.?🤔
CarMoveStrategy 가 Random 을 대체하고 있는데, RacingCarEngine도 반환하게 된다면 내부적으로 random을 수행할 것이고.. 다시 원점이 아닌가 싶어서요~! 저는 오히려 RaceStage#racing이 다른 도메인 객체를 결합 하는 역할을 하고 있다고 생각이 들어요🤔


[getter 없이 로직을 작성해보라는 요구](https://github.com/next-step/kotlin-racingcar/pull/844#discussion_r1022972442)사항은  
 완전히 제거를 해보고 싶었지만, RaceStage가 일급콜렉션의 역할을 가지고 있다 보니 완전히 제거는 어렵더라구요..🤔
사실 RaceStage 에서, Cars로 일급콜렉션을 한번 더 분리해볼까 라는 고민도 해보다가 기능이 단순하여 불필요하다는 생각이 들었는데,
현재 수준에서 리뷰어님은 분리를 하는것을 의도해주신 것인지? 리뷰어 님의 의견이 궁금합니다!
 
마지막까지 잘 부탁드려요 감사합니다.🙇‍♂️

giyeon95 committed:
refactor: winner 조회시 pos 캡슐화

giyeon95 committed:
refactor: 확장함수를 통한 리펙토링

giyeon95 committed:
refactor: 패키지 구조 요구사항에 맞게 변경

ddaaac commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -24,4 +24,8 @@ class Car(
             NON_FORWARD -> this
         }
```
Strategy와 Engine에 관한 코멘트를 남기면서 했던 생각은 두 객체가 나뉘어질 필요가 있을까?하는 생각이었습니다. Engine에서 하고 있는 판단을 Strategy에서 해도 될 것 같아서요. 하나의 클래스로 나타낼 수 있어 보였거든요. 간단하게 나타내면 아래와 같이 나올 것 같네요.

```kotlin
fun race(strategy: Strategy) = if (strategy.canMove()) {
...
```

둘을 합칠 수 있을까라는 얘기를 strategy에서 engine을 리턴한다고 표현했는데 이 부분은 잘못 표현한 것 같네요 😅

여튼 위와 같은 구현에 대한 기연님의 의견이 궁금하네요!

giyeon95 replied:
음... 저는 `RandomMoveStrategy`의 역할이 Random 클래스를 대체하는 방향으로 생각을 했었어요🤔
RacingCarEngine 은 예제코드로 작성해주신 strategy.canMove()를 호출했을때 결국 Boolean 대신 Enum 이 나온다는 점이 다를것 같아요.

다만 작성주신 strategy.canMove() 을 Boolean을 반환하므로 한번에 표현할 수 있겠지만..
이후에 "숫자 중 2가 나오면 후진한다." 라는 요구사항이 생긴다고 가정하고, 확장성을 고려했을때 RacingCarEngine 을 별도로 구현해주는 방식이 조금 더 유연하게 대처할 수 있지 않을까 싶기도 하네요..😁


```kotlin
fun race(strategy: Strategy) = 
when (strategy.runEngine()) {
  FORWARD -> ...
  NON_FORWARD -> ...

}
...

```

결국 두가지를 조합하는 클래스가 RaceStage 가 된다고 생각하였고.. (합치지 않는다면), 저는 다른 역할이 아닐까 싶기도 합니다.🤔
네이밍을 약간 모호하게 지었나 라는 생각도 드네요..🤔🤔

ddaaac replied:
여기도 요구사항의 변화에 대해서 어디까지 생각할까에 대한 얘기가 될 것 같긴 해요.
개인적으로는 요구사항 변화를 예측하는 건 거의 불가능하다고 생각해서 현 요구사항에 맞는 구조만 잡는 게 맞다고 생각합니다. 늘 구체적인 추가 요구사항이 나오면 전혀 예측하지 못한 방향이더라구요. (이건 제가 부족해서일 수도 있구요 🥲)

그래서 여기서 인터페이스를 사용한 이유는 테스트 가능한 구조를 만들기 위해서가 가장 큰 이유이고, 변화에 유연한 구조를 만드는 것과는 조금 거리가 있지 않나 하는 생각을 갖고 있어요. 그래서 현 상황에 맞는 가장 간단한 인터페이스를 도출하는 게 어떨까 하는 게 제 생각이었습니다.

giyeon95 replied:
추가 요구사항이 나오면 예측한 방향과는 다를때가 많다는 점에 공감이 가네요😂
저는 개인적으로 구조에 큰 영향을 주지 않는다면, 약간의 추가 요구사항에 대한 고려를 하는 편인 것 같아요.😁 이유는 미리 고려해두지 않으면 변경 영향도가 커지게 되고, 사이드 이슈도 발생할 수 있다고 생각하기 때문이에요.

다만 가끔보면, 오버 엔지니어링이 되는 경우도 있는것 같기도 하여 @ddaaac 님이 남겨주신 의견도 공감이 가네요.🙏
저도 의견을 공유할 수 있던 시간이라 즐거웠던 것 같습니다.
감사합니다.🙏

ddaaac commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -24,4 +24,8 @@ class Car(
         }
     }
+
+    fun equalsPos(car: Car): Boolean {
+        return this.pos == car.pos
+    }
```
pos를 드러내지 않고 "객체에 메세지를 던져서" 우승자를 구할 수 있지 않을까 하는 생각으로 코멘트를 남겼던 것 같아요. 
조금 극단적으로 우승자를 구하는 조건이 pos가 가장 큰 값이 아닌 이름이 Jason이라면 우승시키기는 조건으로 변한다면 어떻게 될까요? 지금은 이 함수의 이름을 수정하면서 Car와 RaceStage 두 개 클래스의 코드가 모두 수정될 거 같아요.

여기서 pos를 숨기고 조금 일반화된 메세지를 던져볼 수 있으면 Car 클래스만 수정하고도 위 요구사항을 달성할 수 있을 것 같다는 생각이 들었습니다. 여기 함수의 이름을 `isDraw` 등으로 바꾸고, 여기에 `isWin` 등의 함수가 추가되면 어떨까요?
우승자 하나를 찾는 로직은 reduce나 fold를 사용한다면 getter 없이도 구현이 가능할 것 같은데 기연님의 생각은 어떠신가요?

giyeon95 replied:
좋은 사례를 들어주신 것 같아요👍
코멘트 주신 내용이 어떤 의도로 남겨주셨는지 이해가 되었습니다.


getter 없이 구현이 가능하기에, 이전보다 캡슐화가 더 잘 되어있다는 느낌과 reduce가 익숙하지 않음에도 불구하고, 조금 더 코드를 읽기 쉬워보이네요👍

```kotlin
private fun List<Car>.findWinner(): Car = this.reduce { carA, carB -> if (carA.isWinner(carB)) carA else carB }
```
 

```kotlin
private fun List<Car>.findWinner(): Car = this.maxWithOrNull(comparingInt { it.pos }) ?: throw IllegalStateException("우승자가 없을 수 없어요.")
```

ddaaac reviewed (changes requested):
안녕하세요 기연님 👋 

마지막 단계도 잘 구현해주셨어요.

질문 남기신 부분에 대해 코멘트 남겼으니 확인해주세요~

giyeon95 committed:
refactor: pos 캡슐화 리뷰 추가 반영

        pos를 드러내지 않고 "객체에 메세지를 던져서" 우승자를 구할 수 있지 않을까 하는 생각으로 코멘트를 남겼던 것 같아요.
조금 극단적으로 우승자를 구하는 조건이 pos가 가장 큰 값이 아닌 이름이 Jason이라면 우승시키기는 조건으로 변한다면 어떻게 될까요? 지금은 이 함수의 이름을 수정하면서 Car와 RaceStage 두 개 클래스의 코드가 모두 수정될 거 같아요.

여기서 pos를 숨기고 조금 일반화된 메세지를 던져볼 수 있으면 Car 클래스만 수정하고도 위 요구사항을 달성할 수 있을 것 같다는 생각이 들었습니다. 여기 함수의 이름을 `isDraw` 등으로 바꾸고, 여기에 `isWin` 등의 함수가 추가되면 어떨까요?
우승자 하나를 찾는 로직은 reduce나 fold를 사용한다면 getter 없이도 구현이 가능할 것 같은데 기연님의 생각은 어떠신가요?

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/892#discussion_r1025108860_

ddaaac commented on src/main/kotlin/racingcar/domain/RaceStage.kt:
```diff
@@ -22,8 +21,9 @@ class RaceStage(
+        return cars.filter { it.isDraw(winnerCar) }
     }
+
+    private fun List<Car>.findWinner(): Car = this.reduce { carA, carB -> if (carA.isWinner(carB)) carA else carB }
```
reduce 파라미터 람다의 첫번째 인자는 acc, 두번째 인자는 element입니다.
그래서 `{ winner, car -> ~~ }` 요렇게 써주시면 좀 더 의미있는 이름을 사용하실 수 있을 것 같아요.

ddaaac commented on src/test/kotlin/racingcar/domain/CarTest.kt:
```diff
@@ -31,4 +31,38 @@ internal class CarTest {
         // then
         assertThat(afterRaceCar.pos).isEqualTo(expectedPos)
```
만약 getter를 극단적으로 없애고 싶다고 하시면 기존 코드도 isWinner, isDraw를 통해 수정할 수 있습니다. 전진을 했다면 win이고, 하지 않았다면 draw겠죠.
어디까지 고려할까 정도의 차이일 거 같은데, 한 번쯤 고민해 보시면 좋을 것 같아요 😄

giyeon95 replied:
아 테스트코드쪽도 있었네요👍👍
이부분도 고민해볼 수 있겠네요

ddaaac reviewed (approved):
안녕하세요 기연님 👋 

리뷰 너무 잘 반영해주셨네요. 이번 미션은 마무리 지어도 좋을 것 같습니다.

자동차 미션 너무 구현을 잘 해주시고, 코멘트도 꼼꼼히 확인하고 생각 남겨주셔서 리뷰 과정동안 즐거웠습니다 😄 
남은 미션들도 모두 화이팅하시고 즐겁게 진행하시길 바랄게요 🙏
